### PR TITLE
Versioning refactoring and more version information to the log files

### PR DESCRIPTION
**Pull request by bkueng**
in particular, it adds the following information to the logs:
- SW release version (in addition to the git hash)
- OS name & version (tag + git hash if exists)
- mcu version & revision & UUID
- toolchain name & version

Still missing: NuttX version information, see TODO's in version.c. @davids5 can I leave that up to you?

Also I noticed, if you are on a git tag but have local changes, it still shows it as the release version.

### CODE REVIEW

**Comment by kd0aij** [comment]
Thanks, this is very necessary info to have in a log.

File: src/lib/version/version.h [hunk 1]
```diff
@@ -38,33 +38,98 @@
...
+
+/**
+ * Toolchain name used to compile PX4
+ */
+__EXPORT const char *px4_toolchain_name(void);
+
+/**
+ * Toolchain version used to compile PX4 (no particular format)
+ */
+__EXPORT const char *px4_toolchain_version(void);
```
Comment (julianoes) [inline src/lib/version/version.h]: This os going to be useful!

File: src/modules/logger/logger.cpp [hunk 2]
```diff
@@ -1354,9 +1369,46 @@ void Logger::write_header()
...
+	}
+
+	write_info("sys_os_ver_release", px4_os_version());
+	write_info("sys_toolchain", px4_toolchain_name());
+	write_info("sys_toolchain_ver", px4_toolchain_version());
+
+	char rev;
+	char *revstr;
+
+	if (mcu_version(&rev, &revstr) >= 0) {
```
Comment (julianoes) [inline src/modules/logger/logger.cpp]: What's mcu version? I don't have the conext for that. Maybe add a comment?

File: src/modules/logger/params.c [hunk 3]
```diff
@@ -67,3 +67,13 @@ PARAM_DEFINE_INT32(SDLOG_UTC_OFFSET, 0);
  * @group SD Logging
  */
 PARAM_DEFINE_INT32(SDLOG_MODE, 0);
+
+/**
+ * Log UUID
+ *
+ * If set, add an ID to the log, which uniquely identifies the vehicle
```
Comment (julianoes) [inline src/modules/logger/params.c]: But the vehicle has already a UUID from the SoC. Is this one needed on top?

File: src/modules/mavlink/mavlink_messages.cpp [hunk 4]
```diff
@@ -3416,10 +3420,10 @@ class MavlinkStreamAltitude : public MavlinkStream
 			struct home_position_s home;
 			updated |= _home_sub->update(&_home_time, &home);
 
-			if (_global_pos_time > 0 and _home_time > 0) {
+			if (_global_pos_time > 0 && _home_time > 0) {
 				msg.altitude_relative = global_alt - home.alt;
 
-			} else if (_local_pos_time > 0 and _home_time > 0) {
```
Comment (julianoes) [inline src/modules/mavlink/mavlink_messages.cpp]: Huh, how did that compile?

File: src/modules/logger/logger.cpp (hunk 2 above)
Comment (julianoes) [inline src/modules/logger/logger.cpp]: Ah now I see, that's the silicon number.

Comment (bkueng) [inline src/modules/logger/logger.cpp]: I'm changing the variable name. Should be more clear then.

File: src/modules/logger/params.c (hunk 3 above)
Comment (bkueng) [inline src/modules/logger/params.c]: It's only a boolean. I'll rephrase.

Comment (julianoes) [inline src/modules/logger/params.c]: Got it, makes sense.

File: src/modules/mavlink/mavlink_messages.cpp (hunk 4 above)
Comment (LorenzMeier) [inline src/modules/mavlink/mavlink_messages.cpp]: Python!

File: src/modules/uavcannode/uavcannode_main.cpp [hunk 5]
```diff
@@ -238,9 +238,9 @@ void UavcanNode::fill_node_info()
 	/* software version */
 	uavcan::protocol::SoftwareVersion swver;
 
-	// Extracting the first 8 hex digits of FW_GIT and converting them to int
+	// Extracting the first 8 hex digits of the git hash and converting them to int
 	char fw_git_short[9] = {};
-	std::memmove(fw_git_short, FW_GIT, 8);
+	std::memmove(fw_git_short, px4_firmware_version_string(), 8);
 	assert(fw_git_short[8] == '\0');
```
Comment (LorenzMeier) [inline src/modules/uavcannode/uavcannode_main.cpp]: This should be fixed. There must not be a runtime assert in the PX4 codebase.

File: src/modules/mavlink/mavlink_messages.cpp (hunk 4 above)
Comment (kd0aij) [inline src/modules/mavlink/mavlink_messages.cpp]: <%
    if (argc > 1 and argv<:1:> not_eq NULL) <%
        std::cout << "Hello, " << argv<:1:> << '\n';
    %>
%>

File: src/lib/version/version.h [hunk 6]
```diff
@@ -40,36 +40,46 @@
...
 #endif
 
-#endif /* VERSION_H_ */
+
+__BEGIN_DECLS
+
+/**
+ * get the board name as string (including the version if there are multiple)
+ */
+static inline const char *px4_board_name(void)
```
Comment (davids5) [inline src/lib/version/version.h]: @bkueng Did you check the memory map to see the effect of this change?

The direction I wanted to move in is that the board provides the serial number and name via an board_ API.

**Comment by davids5** [comment]
@bkueng 

< Still missing: NuttX version information, see TODO's in version.c. @davids5  can I leave that up to you?

Yep. I have an issue in for a spin on version.

File: src/lib/version/version.h (hunk 6 above)
Comment (bkueng) [inline src/lib/version/version.h]: I checked, the string is only stored once, the function will be reduced to a pointer lookup.

We can do it either way. I just wanted to remove the redundancy between function & marco definition in the board's and to have a well-defined API at one common place.

**Comment by bkueng** [comment]
Thanks @davids5

**Comment by davids5** [comment]
@bkueng https://github.com/PX4/Firmware/pull/7178

**Pull request merged**